executor/cook/launcher.py

```diff
@@ -0,0 +1,181 @@
+                self.process.kill()
+
+        return self.process.returncode
+
+    def poll(self):
+        return self.process.poll()
+
+    def is_guard(self):
+        return self.guard or False
+
+    def is_async(self):
+        return self.async or False
+
+    def is_sync(self):
+        return not self.is_async()
+
+    def is_running(self):
+        return self.process.poll() is None
+
+    def is_crashed(self):
+        return not self.is_running() and self.process.poll() is not 0
+
+def run_commands(commands, stop = None, get_env = lambda: {}, update_task = lambda *_: {}):
+    """
+    Execute a list of commands, waiting for all synchronous commands to complete. Will also
+    exit if the stop event is set.
+
+    Returns a list of exit codes for the command processes.
+    """
+    did_abort = True
+    running_commands = []
+    iterator_commands = (Command(**c) for c in commands)
+
+    try:
+        while not (stop and stop.isSet()):
+            should_await = [c for c in running_commands if c.is_running() and c.is_sync()]
+            should_abort = [c for c in running_commands if c.is_crashed() and c.is_guard()]
+
+            if should_abort:
+                break
+            elif not should_await:
+                if len(commands) == len(running_commands):
+                    did_abort = False
+                    break
+                else:
+                    running_commands.append(next(iterator_commands).run(get_env()))
+            else:
+                time.sleep(1)
+
+            update_task([c.poll() for c in running_commands], False, should_abort)
+    except Exception as e:
+        logging.exception("Exception in run_commands")
+    finally:
+        update_task([c.kill() for c in running_commands], True, did_abort)
+
+        return [c.poll() for c in running_commands]
+
+def munge_commands(task):
+    return task.get('before_commands') + [{'value': task.get('command')}] + task.get('after_commands')
+
+def munge_exit_codes(task, codes):
+    n = len(codes)
+    m = len(task.get('before_commands'))
+
+    d = {}
+
+    if n > m and codes[m] is not None:
+        d.update({'exit_code': codes[m]})
+
+    if n >= m:
+        d.update({'before_exit_codes': codes[0:m]})
+    else:
+        d.update({'before_exit_codes': codes})
+
+    if n > (m + 1):
+        d.update({'after_exit_codes': codes[m + 1:]})
+
+    return d
+
+def run_launcher(store, stop):
+    """
+    Wait for a task to be created and then run its commands. Will also exit if the stop
+    event is set.
+
+    Updates the store with the commands' exit codes upon completion.
+    """
+    try:
+        while not stop.isSet() and not store.all('task'):
+            time.sleep(1)
+
+        if not stop.isSet():
+            id, task = store.all('task').popitem()
+
+            logging.info('Launching task %s', id)
+
+            def get_env():
+                return store.get('task', id).get('env', {})
```

Comment:
Is the environment only set up to include what was explicitly specified in the task configuration?  Should we be setting standard env variables like HOME and USER?